File: lab1/Лаб 1 go/array.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type Array struct {
	Data     []int
	Size     int
	Capacity int
}

func VectorToArray(values []string) *Array {
	arr := &Array{
		Size:     len(values),
		Capacity: len(values),
		Data:     make([]int, len(values)),
	}
	for i, val := range values {
		num, err := strconv.Atoi(val)
		if err == nil {
			arr.Data[i] = num
		} else {
			arr.Data[i] = 0
		}
	}
	return arr
}

func ArrayToVector(arr *Array) []string {
	values := make([]string, arr.Size)
	for i := 0; i < arr.Size; i++ {
		values[i] = strconv.Itoa(arr.Data[i])
	}
	return values
}

func MArrayCreate(values *[]string, sizeStr string) bool {
	size, err := strconv.Atoi(sizeStr)
	if err != nil {
		fmt.Println("Ошибка: неверный формат размера массива")
		return false
	}
	if size < 0 {
		fmt.Println("Ошибка: размер массива не может быть отрицательным")
		return false
	}
	arr := &Array{
		Size:     size,
		Capacity: size,
		Data:     make([]int, size),
	}
	for i := 0; i < arr.Size; i++ {
		arr.Data[i] = 0
	}
	*values = ArrayToVector(arr)
	fmt.Printf("Массив создан, размер: %d\n", size)
	return true
}

func MAdd(values *[]string, value string) bool {
	num, err := strconv.Atoi(value)
	if err != nil {
		fmt.Println("Ошибка: неверный формат числа")
		return false
	}
	arr := VectorToArray(*values)
	if arr.Size >= arr.Capacity {
		arr.Capacity = arr.Capacity * 2
		if arr.Capacity == 0 {
			arr.Capacity = 1
		}
		newData := make([]int, arr.Capacity)
		for i := 0; i < arr.Size; i++ {
			newData[i] = arr.Data[i]
		}
		arr.Data = newData
	}
	arr.Data[arr.Size] = num
	arr.Size++
	*values = ArrayToVector(arr)
	fmt.Printf("Элемент '%s' добавлен в конец массива\n", value)
	return true
}

func MIndexAdd(values *[]string, indexStr, value string) bool {
	index, err := strconv.Atoi(indexStr)
	if err != nil {
		fmt.Println("Ошибка: неверный формат индекса")
		return false
	}
	num, err := strconv.Atoi(value)
	if err != nil {
		fmt.Println("Ошибка: неверный формат числа")
		return false
	}
	arr := VectorToArray(*values)
	if index < 0 || index > arr.Size {
		fmt.Printf("Ошибка: некорректный индекс (%d). Допустимый диапазон: 0-%d\n", index, arr.Size)
		return false
	}
	if arr.Size >= arr.Capacity {
		arr.Capacity = arr.Capacity * 2
		if arr.Capacity == 0 {
			arr.Capacity = 1
		}
		newData := make([]int, arr.Capacity)
		for i := 0; i < arr.Size; i++ {
			newData[i] = arr.Data[i]
		}
		arr.Data = newData
	}
	for i := arr.Size; i > index; i-- {
		arr.Data[i] = arr.Data[i-1]
	}
	arr.Data[index] = num
	arr.Size++
	*values = ArrayToVector(arr)
	fmt.Printf("Элемент '%s' добавлен по индексу %d\n", value, index)
	return true
}

func MGet(values []string, indexStr string) bool {
	index, err := strconv.Atoi(indexStr)
	if err != nil {
		fmt.Println("Ошибка: неверный формат индекса")
		return false
	}
	arr := VectorToArray(values)
	if index < 0 || index >= arr.Size {
		fmt.Printf("Ошибка: некорректный индекс (%d). Допустимый диапазон: 0-%d\n", index, arr.Size-1)
		return false
	}
	fmt.Printf("Элемент по индексу %d: %d\n", index, arr.Data[index])
	return true
}

func MDeleteIndex(values *[]string, indexStr string) bool {
	index, err := strconv.Atoi(indexStr)
	if err != nil {
		fmt.Println("Ошибка: неверный формат индекса")
		return false
	}
	arr := VectorToArray(*values)
	if index < 0 || index >= arr.Size {
		fmt.Printf("Ошибка: некорректный индекс (%d). Допустимый диапазон: 0-%d\n", index, arr.Size-1)
		return false
	}
	removedValue := arr.Data[index]
	for i := index; i < arr.Size-1; i++ {
		arr.Data[i] = arr.Data[i+1]
	}
	arr.Size--
	*values = ArrayToVector(arr)
	fmt.Printf("Элемент '%d' удален по индексу %d\n", removedValue, index)
	return true
}

func MReplace(values *[]string, indexStr, value string) bool {
	index, err := strconv.Atoi(indexStr)
	if err != nil {
		fmt.Println("Ошибка: неверный формат индекса")
		return false
	}
	num, err := strconv.Atoi(value)
	if err != nil {
		fmt.Println("Ошибка: неверный формат числа")
		return false
	}
	arr := VectorToArray(*values)
	if index < 0 || index >= arr.Size {
		fmt.Printf("Ошибка: некорректный индекс (%d). Допустимый диапазон: 0-%d\n", index, arr.Size-1)
		return false
	}
	oldValue := arr.Data[index]
	arr.Data[index] = num
	*values = ArrayToVector(arr)
	fmt.Printf("Элемент по индексу %d заменен: '%d' -> '%s'\n", index, oldValue, value)
	return true
}

func MLength(values []string) bool {
	arr := VectorToArray(values)
	fmt.Printf("Длина массива: %d\n", arr.Size)
	return true
}

File: lab1/Лаб 1/array.cpp
#include "array.h"
#include <iostream>
#include <vector>
#include <string>
#include <sstream>

using namespace std;

Array* vectorToArray(const vector<string>& values) {
    Array* arr = new Array;
    arr->size = values.size();
    arr->capacity = arr->size;
    arr->data = new string[arr->capacity];
    
    for (int i = 0; i < arr->size; i++) {
        arr->data[i] = values[i];
    }
    return arr;
}

void arrayToVector(Array* arr, vector<string>& values) {
    values.clear();
    values.reserve(arr->size);
    
    for (int i = 0; i < arr->size; i++) {
        values.push_back(arr->data[i]);
    }
}

void freeArray(Array* arr) {
    if (arr != nullptr) {
        if (arr->data != nullptr) {
            delete[] arr->data;
        }
        delete arr;
    }
}

bool marrayCreate(vector<string>& values, const string& sizeStr) {
    stringstream ss(sizeStr);
    int size;
    
    if (!(ss >> size) || !ss.eof()) {
        cerr << "Ошибка: неверный формат размера массива" << endl;
        return false;
    }
    
    if (size < 0) {
        cerr << "Ошибка: размер массива не может быть отрицательным" << endl;
        return false;
    }
    
    Array* arr = new Array;
    arr->size = size;
    arr->capacity = size;
    arr->data = new string[arr->capacity];
    
    for (int i = 0; i < arr->size; i++) {
        arr->data[i] = "";
    }
    
    arrayToVector(arr, values);
    freeArray(arr);
    
    cout << "Массив создан, размер: " << size << endl;
    return true;
}

bool madd(vector<string>& values, const string& value) {
    Array* arr = vectorToArray(values);
    
    if (arr->size >= arr->capacity) {
        arr->capacity = (arr->capacity == 0) ? 1 : arr->capacity * 2;
        string* newData = new string[arr->capacity];
        
        for (int i = 0; i < arr->size; i++) {
            newData[i] = arr->data[i];
        }
        
        delete[] arr->data;
        arr->data = newData;
    }
    
    arr->data[arr->size] = value;
    arr->size++;
    
    arrayToVector(arr, values);
    freeArray(arr);
    
    cout << "Элемент '" << value << "' добавлен в конец массива" << endl;
    return true;
}

bool mindexAdd(vector<string>& values, const string& indexStr, const string& value) {
    stringstream ss(indexStr);
    int index;
    
    if (!(ss >> index) || !ss.eof()) {
        cerr << "Ошибка: неверный формат индекса" << endl;
        return false;
    }
    
    Array* arr = vectorToArray(values);
    
    if (index < 0 || index > arr->size) {
        cerr << "Ошибка: некорректный индекс (" << index 
             << "). Допустимый диапазон: 0-" << arr->size << endl;
        freeArray(arr);
        return false;
    }
    
    if (arr->size >= arr->capacity) {
        arr->capacity = (arr->capacity == 0) ? 1 : arr->capacity * 2;
        string* newData = new string[arr->capacity];
        
        for (int i = 0; i < arr->size; i++) {
            newData[i] = arr->data[i];
        }
        
        delete[] arr->data;
        arr->data = newData;
    }
    
    for (int i = arr->size; i > index; i--) {
        arr->data[i] = arr->data[i - 1];
    }
    
    arr->data[index] = value;
    arr->size++;
    
    arrayToVector(arr, values);
    freeArray(arr);
    
    cout << "Элемент '" << value << "' добавлен по индексу " << index << endl;
    return true;
}

bool mget(const vector<string>& values, const string& indexStr) {
    stringstream ss(indexStr);
    int index;
    
    if (!(ss >> index) || !ss.eof()) {
        cerr << "Ошибка: неверный формат индекса" << endl;
        return false;
    }
    
    Array* arr = vectorToArray(values);
    
    if (index < 0 || index >= arr->size) {
        cerr << "Ошибка: некорректный индекс (" << index 
             << "). Допустимый диапазон: 0-" << (arr->size - 1) << endl;
        freeArray(arr);
        return false;
    }
    
    cout << "Элемент по индексу " << index << ": " << arr->data[index] << endl;
    freeArray(arr);
    return true;
}

bool mdeleteIndex(vector<string>& values, const string& indexStr) {
    stringstream ss(indexStr);
    int index;
    
    if (!(ss >> index) || !ss.eof()) {
        cerr << "Ошибка: неверный формат индекса" << endl;
        return false;
    }
    
    Array* arr = vectorToArray(values);
    
    if (index < 0 || index >= arr->size) {
        cerr << "Ошибка: некорректный индекс (" << index 
             << "). Допустимый диапазон: 0-" << (arr->size - 1) << endl;
        freeArray(arr);
        return false;
    }
    
    string removedValue = arr->data[index];
    
    for (int i = index; i < arr->size - 1; i++) {
        arr->data[i] = arr->data[i + 1];
    }
    
    arr->size--;
    
    arrayToVector(arr, values);
    freeArray(arr);
    
    cout << "Элемент '" << removedValue << "' удален по индексу " << index << endl;
    return true;
}

bool mreplace(vector<string>& values, const string& indexStr, const string& value) {
    stringstream ss(indexStr);
    int index;
    
    if (!(ss >> index) || !ss.eof()) {
        cerr << "Ошибка: неверный формат индекса" << endl;
        return false;
    }
    
    Array* arr = vectorToArray(values);
    
    if (index < 0 || index >= arr->size) {
        cerr << "Ошибка: некорректный индекс (" << index 
             << "). Допустимый диапазон: 0-" << (arr->size - 1) << endl;
        freeArray(arr);
        return false;
    }
    
    string oldValue = arr->data[index];
    arr->data[index] = value;
    
    arrayToVector(arr, values);
    freeArray(arr);
    
    cout << "Элемент по индексу " << index << " заменен: '" 
         << oldValue << "' -> '" << value << "'" << endl;
    return true;
}

bool mlength(const vector<string>& values) {
    Array* arr = vectorToArray(values);
    cout << "Длина массива: " << arr->size << endl;
    freeArray(arr);
    return true;
}
